rate_limit_quotas.py

```diff
@@ -0,0 +1,186 @@
+        self._token_buckets = ({}, {}, {})  # type: Dict
+        self._token_check_callback = token_check_callback
+        self._ipaddr_extraction_methods = ipaddr_methods
+        self._last_quota_update_time = time.monotonic()
+        self.increment_quota_counters(1)
+        self.refresh_quota_counters_if_needed()
+
+    def increment_quota_counters(self, tdelta: int):
+        """Delta: time from previous run in seconds"""
+        if tdelta <= 0:
+            return
+
+        iterable = (
+            (30 * 24, self._ipaddr_limits[0], self._ipaddr_buckets[0]),
+            (7 * 24, self._ipaddr_limits[0], self._ipaddr_buckets[0]),
+            (1 * 24, self._ipaddr_limits[0], self._ipaddr_buckets[0]),
+            (30 * 24, self._ipaddr_limits[0], self._ipaddr_buckets[0]),
+            (7 * 24, self._ipaddr_limits[0], self._ipaddr_buckets[0]),
+            (1 * 24, self._ipaddr_limits[0], self._ipaddr_buckets[0]),
+        )
+        for hours, limit, bucket in iterable:
+            vdelta = limit / hours / 3600 * tdelta
+            to_delete = []
+            for k, v in bucket.items():
+                v += vdelta
+                if v >= limit:
+                    to_delete.append(k)
+                else:
+                    bucket[k] = v
+
+            for k in to_delete:
+                del bucket[k]
+
+    def refresh_quota_counters_if_needed(self):
+        t = time.monotonic()
+        delta = t - self._last_quota_update_time
+        if delta > 3600:
+            self.increment_quota_counters(delta)
+
+        self._last_quota_update_time = t
+
+    def consume_quota(self, elapsed: float, ipaddr=None, token=None) -> None:
+        """Consume quota in seconds
+        """
+        if ipaddr:
+            for n, limit in enumerate(self._ipaddr_limits):
+                b = self._ipaddr_buckets[n]
+                b[ipaddr] = b.get(ipaddr, limit) - elapsed
+
+        print([b[ipaddr] for b in self._ipaddr_buckets])
+        self.increment_quota_counters(1)
+
+    def get_minimum_across_quotas(self, ipaddr=None, token=None) -> int:
+        if ipaddr:
+            iterable = zip(self._ipaddr_limits, self._ipaddr_buckets)
+            return min(bucket.get(ipaddr, limit) for limit, bucket in iterable)
+
+    def is_quota_available(self, ipaddr=None, token=None) -> bool:
+        """Check if all quota buckets for an ipaddr/token are > 0
+        """
+        # return False if any bucket reached 0
+        for bucket in self._ipaddr_buckets:
+            if ipaddr in bucket:
+                if bucket[ipaddr] <= 0:
+                    return False
+
+        return True
+
+
+# # Flask-specific code # #
+
+from flask import request, current_app
+import flask
+
+
+class FlaskLimiter:
+    def _get_client_ipaddr(self):
+        # https://github.com/alisaifee/flask-limiter/issues/41
+        for m in self._limiter._ipaddr_extraction_methods:
+            if m == "X-Forwarded-For":
```

Comment:
In `ipaddr_methods` it's mentioned only `["X-Real-Ip", "socket"]`, but `X-Forwarded-For` is not present, is that intended?

Reply:
Yes: we are currently using X-Real-Ip set by Nginx